crypto/s390xcap.c

```diff
@@ -13,39 +13,52 @@
 #include <setjmp.h>
 #include <signal.h>
 #include "internal/cryptlib.h"
-
-extern unsigned long OPENSSL_s390xcap_P[];
+#include "s390x_arch.h"
 
 static sigjmp_buf ill_jmp;
 static void ill_handler(int sig)
 {
     siglongjmp(ill_jmp, sig);
 }
 
-unsigned long OPENSSL_s390x_facilities(void);
+void OPENSSL_s390x_facilities(void);
+void OPENSSL_vx_probe(void);
 
 void OPENSSL_cpuid_setup(void)
 {
     sigset_t oset;
     struct sigaction ill_act, oact;
 
-    if (OPENSSL_s390xcap_P[0])
+    if (OPENSSL_s390xcap_P.stfle[0])
         return;
 
-    OPENSSL_s390xcap_P[0] = 1UL << (8 * sizeof(unsigned long) - 1);
+    OPENSSL_s390xcap_P.stfle[0] = 1ULL << (sizeof(OPENSSL_s390xcap_P.stfle[0])
+                                           * 8 - 1);
 
     memset(&ill_act, 0, sizeof(ill_act));
     ill_act.sa_handler = ill_handler;
     sigfillset(&ill_act.sa_mask);
     sigdelset(&ill_act.sa_mask, SIGILL);
+    sigdelset(&ill_act.sa_mask, SIGFPE);
     sigdelset(&ill_act.sa_mask, SIGTRAP);
     sigprocmask(SIG_SETMASK, &ill_act.sa_mask, &oset);
     sigaction(SIGILL, &ill_act, &oact);
+    sigaction(SIGFPE, &ill_act, &oact);
 
     /* protection against missing store-facility-list-extended */
     if (sigsetjmp(ill_jmp, 1) == 0)
         OPENSSL_s390x_facilities();
 
+    /* protection against disabled vector facility */
+    if (sigsetjmp(ill_jmp, 1) == 0)
```

Comment:
This should be in above if-clause. I mean if OPENSSL_s390x_facilities() causes SIGILL (or equivalent), then there is no point probing for VX. And it has to have additional if checking if any of VX capability bit is set. I mean the purpose of the probe is to *mask* the bits if kernel doesn't engage vector register bank, right? And there is no point in probing VX if bits were not there to start with.

Reply:
It should be noted that instead of placing the VX probe in above if, one can as well just add that second if alone. Indeed, if OPENSSL_s390x_facilities has failed, then VX bits will be clear. I actually *might* be more readable that way.... I mean think about it...

Reply:
if the facilities function fails, the vx bit is clear. I would prepend a check for the vx bit in the second "if". I would not nest the "if"s. Are you okay with this ?

Reply:
on second thought checking the vx bit is not really needed. on machines w/o vx, it will just throgh a SIGILL.

Reply:
The SIGILL probes have an annoying side effect: when you run application under debugger it will break on failed probes. For this reason it's deemed appropriate to *minimize* SIGILL probes whenever possible. For this reason suggested snippet is preferred. Modulo style nit, `{}` should be in both `if` and `else` or neither.